### PR TITLE
Enhancement: Custom FullCalendar Configuration for Widgets

### PR DESCRIPTION
**Pull request by sitenzo**
In this pull request, we're introducing a feature that allows you to customize the FullCalendar configuration on a per-widget basis.

### What's New?
You can now set a default configuration at the plugin level. However, if you need to make specific changes for a particular widget, you have the flexibility to do so.

### How to Customize Widget Configuration?
In your widget, implement the getConfig method.

```
public function getConfig(): array {
    return [
        // Your custom FullCalendar configuration settings go here
    ];
}
```

Set up the custom configuration for the widget within the getConfig method.

### Merging Configurations
Behind the scenes, both the default plugin configuration and the widget-specific configuration files are merged into one final configuration for the widget. This ensures that you have full control over how FullCalendar behaves for each widget.

We believe this enhancement will provide greater flexibility and customization options for your FullCalendar widgets.

### CODE REVIEW

**Comment by saade** [comment]
Hey! Thanks for your contribution, i'll review it soon :)

**Comment by StaffCollab** [comment]
Hello.
This is a very needed feature! 
I want to customize the hidden days and times per user to only show the available times to book sessions on a per-caregiver availability basis.

This would be wonderful.

~~~
    protected function getFooterWidgets(): array
    {
        return [
            CalendarWidget::make()
                ->config([
                    'hiddenDays' => $this->record->notAvailableDays,
                ]),
        ];
    }
~~~

**Comment by lukas-frey** [comment]
Any news on this PR? :)

**Comment by GeoSot** [comment]
@saade can we help, in order to merge this?

File: resources/views/fullcalendar.blade.php [hunk 1]
```diff
@@ -16,7 +16,7 @@
                 plugins: @js($plugin->getPlugins()),
                 schedulerLicenseKey: @js($plugin->getSchedulerLicenseKey()),
                 timeZone: @js($plugin->getTimezone()),
-                config: @js($plugin->getConfig()),
+                config: @js($this->getFinalConfig()),
```
Comment (saade) [inline resources/views/fullcalendar.blade.php]:
```suggestion
                config: @js($this->getConfig()),
```

File: src/Widgets/Concerns/InteractsWithConfig.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Saade\FilamentFullCalendar\Widgets\Concerns;
+
+use Saade\FilamentFullCalendar\FilamentFullCalendarPlugin;
+
+trait InteractsWithConfig
```
Comment (saade) [inline src/Widgets/Concerns/InteractsWithConfig.php]:
```suggestion
trait CanBeConfigured
```

File: src/Widgets/Concerns/InteractsWithConfig.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+{  
+	public function getFinalConfig()
```
Comment (saade) [inline src/Widgets/Concerns/InteractsWithConfig.php]:
```suggestion
	public function config(): array
```

File: README.md [hunk 4]
```diff
@@ -218,6 +218,31 @@ class AdminPanelProvider extends PanelProvider
...
+namespace App\Filament\Widgets;
+
+use Saade\FilamentFullCalendar\Widgets\FullCalendarWidget;
+use App\Models\Event;
+
+class CalendarWidget extends FullCalendarWidget
+{
+    public Model | string | null $model = Event::class;
+
+    public function getConfig(): array
```
Comment (saade) [inline README.md]:
```suggestion
    public function config(): array
```

File: src/Widgets/Concerns/InteractsWithConfig.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	{
+		$initConfig = FilamentFullCalendarPlugin::get()->getConfig();
+		$widgetConfig = $this->getConfig();
+		return self::mergeArraysRecursive($initConfig,$widgetConfig);
```
Comment (saade) [inline src/Widgets/Concerns/InteractsWithConfig.php]:
```suggestion
		return [];
```

File: src/Widgets/Concerns/InteractsWithConfig.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+			if (is_array($value) && isset($initial[$key]) && is_array($initial[$key])) {
+				// If both are arrays, recursively merge them
+				$initial[$key] = mergeArraysRecursive($initial[$key], $value);
+			} else {
+				// Otherwise, overwrite the value
+				$initial[$key] = $value;
+			}
+		}
+		return $initial;
+	}
```
Comment (saade) [inline src/Widgets/Concerns/InteractsWithConfig.php]:
```suggestion
```

File: src/Widgets/Concerns/InteractsWithConfig.php [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+
+	  protected function getConfig(): array
+    {
+        return [];
+    }
```
Comment (saade) [inline src/Widgets/Concerns/InteractsWithConfig.php]:
```suggestion
    protected function getConfig(): array
    {
        return array_merge_recursive(
            FilamentFullCalendarPlugin::get()->getConfig(),
            $this->config(),
        );
    }
```

File: src/Widgets/FullCalendarWidget.php [hunk 8]
```diff
@@ -18,7 +18,8 @@ class FullCalendarWidget extends Widget implements HasForms, HasActions
     use Concerns\InteractsWithRecords;
     use Concerns\InteractsWithHeaderActions;
     use Concerns\InteractsWithModalActions;
-
+    use Concerns\InteractsWithConfig;
```
Comment (saade) [inline src/Widgets/FullCalendarWidget.php]:
```suggestion
    use Concerns\CanBeConfigured;
```

File: src/Widgets/FullCalendarWidget.php [hunk 9]
```diff
@@ -57,4 +58,9 @@ public function getFormSchema(): array
     {
         return [];
     }
+
+    public function getConfig(): array
+	{
+		return [];
+	}
```
Comment (saade) [inline src/Widgets/FullCalendarWidget.php]:
```suggestion
```

**Review by saade — CHANGES_REQUESTED** [review]
Hey! Sorry for the low support on this plugin. I have big improvements for it on my TODO but i'm busy in another project atm.

If someone could apply and test those changes and test i'll be happy to merge. Thanks!

**Comment by GeoSot** [comment]
I am not sure if any other than @sitenzo can edit this PR

**Comment by saade** [comment]
I'll have a look later today 👍🏼

**Comment by sitenzo** [comment]
I have committed all the changes you have reviewed.

**Comment by GeoSot** [comment]
[array_merge_recursive](https://www.php.net/manual/en/function.array-merge-recursive.php), maybe is not the perfect solution here 
_Possible help: https://copyprogramming.com/howto/php-merge-arrays-with-only-not-duplicated-values_


Edit: @saade, supposing you bing back [`mergeArraysRecursive`](https://github.com/saade/filament-fullcalendar/pull/130/commits/536cf81eec33021ed7d467ced44178e76b4b6e48), from the initial commit, it will be fine to go

**Comment by Sjoerd24** [comment]
Any news on this pull request? This option would be extremely useful for me :) I think with the suggestion of @GeoSot to bring back [mergeArraysRecursive](https://github.com/saade/filament-fullcalendar/pull/130/commits/536cf81eec33021ed7d467ced44178e76b4b6e48) the PR should be fine right? Looks good to my eye at least 👍

**Review by saade — APPROVED** [review]

**Comment by saade** [comment]
Thanks!

**Pull request merged**